[MarkdownFixer] Do not escape username with underscores in code or codeblocks

JuanitoFatas opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [x] Tests are passing 

  ```shell
  bundle exec rspec spec/labor/markdown_fixer_spec.rb spec/requests/api/v0/articles_spec.rb spec/services/articles/creator_spec.rb spec/services/mentions/create_all_spec.rb
  ```
- [x] I‘ve verified it works on development
  starts the server at local, sign in, new post, copy the following to preview and see if things work as expected. You should not see `\` in code and codeblock, and underscored username renders correctly.
  <details>
  <summary>Code for testing at development</summary>

  <pre>
  @_dev_

  ```
  @_dev_
  ```

  `@_dev_`

  @_dev_
  </pre>
  </details>

## Description

The underscored username got escaped in code and code blocks. We probably do not want that. The fix is only to escape underscored username if they are not in code or code blocks.

Other possible approaches:

- Write regexp that only pull out underscored usernames not in code or code blocks
- Turn markdown into HTML and use Nokogiri to work on text of ancestors not `pre` or `code`

## Related Tickets & Documents

Fixes #5739.  

## Screenshots

### Current on dev (2019.01.28, 26e4412)

<img width="466" alt="スクリーンショット 2020-01-28 17 41 26" src="https://user-images.githubusercontent.com/1000669/73248000-755c5600-41f5-11ea-83bf-cbae11a8e0f3.png">

### After this Pull Request (tested against 432885f)

<img width="260" alt="スクリーンショット 2020-02-09 22 49 42" src="https://user-images.githubusercontent.com/1000669/74103332-98d8b680-4b8e-11ea-99b7-e4c0aff80a28.png">

<img width="187" alt="スクリーンショット 2020-02-09 22 36 07" src="https://user-images.githubusercontent.com/1000669/74103163-3632eb00-4b8d-11ea-837b-02fa662cf22c.png">

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed
- [x] code comments added

## [optional] What gif best describes this PR or how it makes you feel?

![](https://media.giphy.com/media/Vuw9m5wXviFIQ/giphy.gif)

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=5785) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=5785) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=5785) it.</sub>

Zhao-Andy commented:
I assume this PR is good because all tests are passing now, but @JuanitoFatas let me know if I'm mistaken. :)

JuanitoFatas commented:
@Zhao-Andy 

> I assume this PR is good because all tests are passing now, but @JuanitoFatas let me know if I'm mistaken. :)

Rebased with master and verified it works ✅ 

<img width="290" alt="スクリーンショット 2020-02-04 09 27 52" src="https://user-images.githubusercontent.com/1000669/73702743-0e78f880-4731-11ea-863b-9c17c8298874.png">

and tests passed 🎉 

<img width="1418" alt="スクリーンショット 2020-02-04 09 32 22" src="https://user-images.githubusercontent.com/1000669/73702801-42ecb480-4731-11ea-839f-4808f562e38c.png">

:shipit:

rhymes commented on app/labor/markdown_fixer.rb:
```diff
@@ -47,7 +68,28 @@ def split_tags(markdown)
+      end.join
+    end
+
+    USERNAME_WITH_UNDERSCORE_REGEXP = /(?<!`)@_\w+_/.freeze
```
can we add a comment that narrates this regexp? As it's not particularly intuitive

JuanitoFatas commented:
@rhymes 

Thanks for the review. I‘ve added comments and changed the implementation to a less clever one 😉 . I also pulled the node traverser into a separate file, hope that make it easier to read. Verified this works and updated the screenshot in the Pull Request accordingly 🎉

citizen428 commented on app/labor/markdown_fixer.rb:
```diff
@@ -47,11 +47,29 @@ def split_tags(markdown)
 
     def underscores_in_usernames(markdown)
-      markdown.gsub(/(@)(_\w+)(_)/, '\1\\\\\2\\\\\3')
+      return markdown unless markdown.include?("@")
```
This guard clause has to go through all of `markdown` to determine whether or not to return, so I don't think it saves us much work. If we skip this check we go through the document once, if we keep it we go through the document once + the percentage of text preceding the username.

JuanitoFatas replied:
There are number of tests that pass `""`, front matters, or text that do not have username at all (without `@`) to here. These make sense in real world. Hence checking for `@` here to avoid doing work for irrelevant cases and all tests are happy.

citizen428 replied:
I understand that the check makes sense, but I think the way it's currently done is slightly inefficient. You already walk the document in the traverser, you could just ignore lines without an `@`, so you only walk the document a total of 1 time. Was that a bit clearer?

citizen428 replied:
I understand the reason for the check, my only slight concern was that in articles that contain usernames we actually do slightly more work (first read `markdown` until we find an `@`, then traverse it), but given that the majority of articles will probably not have usernames in them this seems like a good solution.

JuanitoFatas replied:
Hey checking against the regexp makes sense. I‘ve updated the code.

citizen428 commented on app/labor/markdown_fixer.rb:
```diff
@@ -47,11 +47,29 @@ def split_tags(markdown)
+
+      traverser = MarkdownTraverser.new(markdown)
+      traverser.each do |line|
+        unless traverser.in_codeblock?
```
How about 

```ruby
next if traverser.in_codeblock?
escape_underscored_username_in_line!(line)
```

JuanitoFatas replied:
Sure 👍

citizen428 commented on app/labor/markdown_fixer.rb:
```diff
@@ -47,11 +47,29 @@ def split_tags(markdown)
+
+    # Escapes underscored username that is not in code
+    def escape_underscored_username_in_line!(line)
+      line.scan(USERNAME_WITH_UNDERSCORE_REGEXP).each do |to_escape|
```
Instead of `scan` + `each` and then explicitly returning `line`, could we use the block form of `gsub` here? Something like (untested)

```ruby
line.gsub(USERNAME_WITH_UNDERSCORE_REGEXP) { |to_escape| to_escape.gsub("_", "\\_") }
```

JuanitoFatas replied:
I would like this too, but the `line` could be

```text
@_dev_ `@_dev_`
```

hence we cannot replace all underscores in a line.

citizen428 replied:
This is not replacing all underscores in a line, this is replacing all underscores in  every `to_escape`, which is everything that matches `USERNAME_WITH_UNDERSCORE_REGEXP`:

```ruby
line = '@_dev_ `@_dev_`'
line.gsub(USERNAME_WITH_UNDERSCORE_REGEXP) { |to_escape| to_escape.gsub("_", "\\_") }
#=> "@\\_dev\\_ `@_dev_`"
```

Isn't this what you want, i.e. escape the username outside backticks but leave the one inside unchanged?

JuanitoFatas replied:
Sorry I overlooked from my last comment. But I did give this a try and other places need to modify as well but I don‘t have time to look into why this does not work. What‘s your concern on the current implementation? I‘m happy to receive a PR against this feature branch.

citizen428 replied:
No concern, it just seemed like it's using a more complicated approach than necessary. I'll trust you if you say it doesn't work, so let's just leave it as is.

citizen428 commented on app/labor/markdown_traverser.rb:
````diff
@@ -0,0 +1,35 @@
+
+  private
+
+  CODEBLOCK_MARKER = "```".freeze
````
Not that the visibility of this matters much, just checking in to see if you're aware that moving constants in a `private` block will not actually make them private, you need the `private_constant` method for that.

JuanitoFatas replied:
Yup 👍
